### PR TITLE
ao/co: Report scan progress for CREATE INDEX

### PR DESCRIPTION
**Pull request by soumyadeep2007**
Please see individual commits for details.

### CODE REVIEW

File: src/include/cdb/cdbappendonlyam.h [hunk 1]
```diff
@@ -237,6 +237,12 @@ typedef struct AppendOnlyScanDescData
 	int			rs_cindex;		/* current tuple's index in tbmres->offsets */
 	struct AppendOnlyFetchDescData *aofetch;
 
+	/*
+	 * The total number of bytes read, uncompressed, across all segment files,
```
Comment (ashwinstar) [inline src/include/cdb/cdbappendonlyam.h:241]: s/uncompressed/compressed

File: src/include/cdb/cdbappendonlyam.h [hunk 2]
```diff
@@ -443,4 +449,20 @@ extern TM_Result appendonly_delete(
...
+ * update it with the compressed length. If the block was not compressed, update
+ * it with the uncompressed length.
+ */
+static inline void
+AppendOnlyScanDesc_UpdateTotalBytesRead(AppendOnlyScanDesc scan)
+{
+	Assert(scan->storageRead.isActive);
+
+	if (scan->storageRead.current.isCompressed)
+		scan->totalBytesRead += scan->storageRead.current.compressedLen;
```
Comment (ashwinstar) [inline src/include/cdb/cdbappendonlyam.h:463]: I thought compressedLen provided correct value in both cases, maybe not

File: src/test/isolation2/expected/ao_index_build_progress.out [hunk 3]
```diff
@@ -0,0 +1,49 @@
+-- Test to ensure we correctly report progress in pg_stat_progress_create_index
+-- for append-optimized tables
+
+-- AO table
+CREATE TABLE ao_index_build_progress(i int, j bigint) USING ao_row WITH (compresstype=zlib, compresslevel=2);
```
Comment (ashwinstar) [inline src/test/isolation2/expected/ao_index_build_progress.out:5]: prefer we use zstd in most of tests needing compression as that's the recommended method hence testing with it more would be good

**Review by ashwinstar — COMMENTED** [review]
So should be document `tuples_total` and `tuples_done` will not be reported for AO/CO tables? or we have plans to even report that in-future?

File: src/test/isolation2/expected/ao_index_build_progress.out [hunk 4]
```diff
@@ -47,3 +47,50 @@ SELECT gp_inject_fault('AppendOnlyStorageRead_ReadNextBlock_success', 'reset', d
...
+-------------------------------
+ Success:                      
+(1 row)
+
+-- By now, we should have reported some blocks (of size 'block_size') as "done",
+-- as well as a total number of blocks that matches the relation's on-disk size.
+1U: SELECT command, phase, (pg_relation_size('aoco_index_build_progress') + (current_setting('block_size')::int - 1)) / current_setting('block_size')::int AS blocks_total_actual, blocks_total AS blocks_total_reported, blocks_done AS blocks_done_reported FROM pg_stat_progress_create_index WHERE relid = 'aoco_index_build_progress'::regclass;
+ command      | phase                          | blocks_total_actual | blocks_total_reported | blocks_done_reported 
+--------------+--------------------------------+---------------------+-----------------------+----------------------
+ CREATE INDEX | building index: scanning table | 5                   | 5                     | 3
```
Comment (ashwinstar) [inline src/test/isolation2/expected/ao_index_build_progress.out:86]: For CO tables, for first index creation all column segfiles are read as need to build block directory. For all following create index only indexed columns are read. Does the current code take that into account and report progress based on that? Need test to validate that aspect. (not sure we have interface to get just size of a column, might have to use EOF from pg_aocsseg)

**Comment by soumyadeep2007** [comment]
> So should be document tuples_total and tuples_done will not be reported for AO/CO tables? or we have plans to even report that in-future?

Let me clarify as this threw me off too and is a little confusing (its horrible upstream code - an absolute jungle!!):

* There are phases and subphases (index AM dependent) within the "building index" phase. Also see `ambuildphasename`.
* During the `PROGRESS_BTREE_PHASE_INDEXBUILD_TABLESCAN` subphase blocks are reported, whereas in the 
In the PROGRESS_BTREE_PHASE_INDEXBUILD_TABLESCAN "subphase", blocks are reported (from the table AM code), while the table is scanned. Then in the PROGRESS_BTREE_PHASE_LEAF_LOAD subphase, tuples are actually reported, as they are "loaded" from the spool (`loading tuples in tree` subphase). Hash indexes are similar too.
* One reason why tuples are reported only in the index load phase is it would be incredibly expensive to report by tuple in the scan subphase for reasonably sized relations.

File: src/include/cdb/cdbappendonlyam.h (hunk 1 above)
Comment (soumyadeep2007) [reply src/include/cdb/cdbappendonlyam.h:241]: Thanks for the catch!

File: src/test/isolation2/expected/ao_index_build_progress.out (hunk 3 above)
Comment (soumyadeep2007) [reply src/test/isolation2/expected/ao_index_build_progress.out:5]: Hmmm none of our tests currently use zstd, and I was worried about whether CI builds with it etc. I can try, thanks for pointing this out!

File: src/include/cdb/cdbappendonlyam.h (hunk 2 above)
Comment (soumyadeep2007) [reply src/include/cdb/cdbappendonlyam.h:463]: Nope. See for eg:
```
AppendOnlyStorageFormat_GetSmallContentHeaderInfo()
...
*compressedLen = AOSmallContentHeaderGet_compressedLength(blockHeader);
if (*compressedLen == 0)
{
	*isCompressed = false;
	length = *uncompressedLen;
}
else
{
	*isCompressed = true;
	length = *compressedLen;
```

File: src/test/isolation2/expected/ao_index_build_progress.out (hunk 4 above)
Comment (soumyadeep2007) [reply src/test/isolation2/expected/ao_index_build_progress.out:86]: Yes, I was wondering if create index considered col projection! Thanks for the pointer, all we have to do is make the total blocks calculation aware of projection info. I will add a test too!

Reference:
```
/*
 * if block directory has created, we can only scan needed column.
 */
```

File: src/test/isolation2/expected/ao_index_build_progress.out (hunk 3 above)
Comment (soumyadeep2007) [reply src/test/isolation2/expected/ao_index_build_progress.out:5]: We don't build with zstd in our pipelines unfortunately, so I will stick w/ zlib for this PR.

File: src/test/isolation2/expected/ao_index_build_progress.out (hunk 4 above)
Comment (soumyadeep2007) [reply src/test/isolation2/expected/ao_index_build_progress.out]: Done!

File: src/test/isolation2/expected/ao_index_build_progress.out (hunk 3 above)
Comment (ashwinstar) [reply src/test/isolation2/expected/ao_index_build_progress.out:5]: That doesn't feel right think to do, we need to fix it.

Comment (soumyadeep2007) [reply src/test/isolation2/expected/ao_index_build_progress.out]: Done in this PR itself as a separate commit, as the CI change was a one-liner.

**Comment by soumyadeep2007** [comment]
Removed the compression ratio output from the tests as there are subtle variations in CI such as the following (probably due to libzstd version differences):

```diff
diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /tmp/build/629a37e4/gpdb_src/src/test/isolation2/expected/ao_index_build_progress.out /tmp/build/629a37e4/gpdb_src/src/test/isolation2/results/ao_index_build_progress.out
--- /tmp/build/629a37e4/gpdb_src/src/test/isolation2/expected/ao_index_build_progress.out	2022-12-30 02:19:52.933231651 +0000
+++ /tmp/build/629a37e4/gpdb_src/src/test/isolation2/results/ao_index_build_progress.out	2022-12-30 02:19:52.941232482 +0000
@@ -13,7 +13,7 @@
 SELECT get_ao_compression_ratio('ao_index_build_progress'::regclass);
  get_ao_compression_ratio 
 ------------
- 8.8                      
+ 8.81                     
 (1 row)
```

**Review by bmdoil — APPROVED** [review]
Tested and validated the implementation. 

This will be great on the customer side for investigating how long running CREATE INDEX statements are progressing, but is it possible to enable reporting from the coordinator? Connecting in U mode to a segment positions this as more of a developer debugging tool because we really should be discouraging end users connecting in U mode. 

Also, the `building index: sorting live tuples` took awhile in my test, not as long as the table scan but still significant. Is there any benefit to progress reporting that?

Both of these comments are outside the scope of this PR, so it LGTM.

**Pull request merged**